storage/simulation/range.go

```diff
@@ -116,10 +128,59 @@ func (r *Range) splitRange(originalRange *Range) {
 	stores := originalRange.getStores()
 	r.Lock()
 	defer r.Unlock()
-	r.desc.Replicas = desc.Replicas
+	r.desc.Replicas = append([]proto.Replica(nil), desc.Replicas...)
 	r.stores = stores
 }
 
+// prepareActions walks through each replica and determines if any action is
+// required using the allocator.
+func (r *Range) prepareActions() {
+	r.Lock()
+	defer r.Unlock()
+	r.nextActions = make(map[proto.StoreID]nextAction)
+	for storeID := range r.stores {
+		action, priority := r.allocator.ComputeAction(r.zone, &r.desc)
+		var rebalance bool
+		if action == storage.AANoop {
+			rebalance = r.allocator.ShouldRebalance(storeID)
+			priority = 0
+		}
+		r.nextActions[storeID] = nextAction{
+			action:    action,
+			priority:  priority,
+			rebalance: rebalance,
+		}
+	}
+}
+
+// getNextAction returns a single action based on the priorities of the each
+// action stored in nextActions.
+func (r *Range) getNextAction() nextAction {
+	r.RLock()
+	defer r.RUnlock()
+	var topAction nextAction
+	// TODO(bram): This is random. Might want to make it deterministic for
+	// repeatability.
+	for _, nextAction := range r.nextActions {
+		if nextAction.priority > topAction.priority {
+			topAction = nextAction
+		}
+	}
+	return topAction
+}
+
+// getAllocateTarget calls allocateTarget for the range and returns the top
+// target store.
+func (r *Range) getAllocateTarget() proto.StoreID {
+	r.RLock()
```

Comment:
the use of this mutex is pretty inconsistent in this file. For instance, take a look at `(*Range).splitRange(*Range)`: that function acquires the "other" lock twice.

I think the `getFoo` methods in this file are more trouble than they're worth, but I'm open to a counter-argument.